Seata-XA not working on Oracle

- [x] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
Seata-xa mode can't work on Oracle 
oracle version: Database 11g Enterprise Edition Release 11.2.0.4.0 - 64bit Production
project: https://github.com/seata/seata-samples/tree/master/seata-xa

### Attention
It's one Oracle Instance not multi oracle instance. 
All Table in the same oracle schema.

### Ⅱ. Describe what happened
When access http://localhost:8084/purchase , the page output is: Purchase Failed:订单服务调用失败,事务回滚!
 
The console output stacktrace as below:

```
2020-12-18 17:25:10.595  INFO 18272 --- [nio-8082-exec-1] io.seata.sample.service.OrderService     : create order in transaction: 172.17.0.1:8091:83249848442163200
2020-12-18 17:25:10.595  INFO 18272 --- [nio-8082-exec-1] io.seata.sample.service.OrderService     : sql insert into order_tbl(id, user_id,commodity_code,count,money) values(1,'U100000','C100000',30,3000)
org.springframework.jdbc.UncategorizedSQLException: StatementCallback; uncategorized SQLException for SQL [insert into order_tbl(id, user_id,commodity_code,count,money) values(1,'U100000','C100000',30,3000)]; SQL state [null]; error code [0]; failed to start xa branch 172.17.0.1:8091:83249848442163200 since null; nested exception is java.sql.SQLException: failed to start xa branch 172.17.0.1:8091:83249848442163200 since null
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:89)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81)
	at org.springframework.jdbc.core.JdbcTemplate.translateException(JdbcTemplate.java:1414)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:388)
	at org.springframework.jdbc.core.JdbcTemplate.update(JdbcTemplate.java:523)
	at io.seata.sample.service.OrderService.create(OrderService.java:40)
	at io.seata.sample.controller.OrderController.create(OrderController.java:21)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:215)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:142)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:102)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:800)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1038)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:942)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:998)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:890)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:634)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:875)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at com.alibaba.druid.support.http.WebStatFilter.doFilter(WebStatFilter.java:123)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:92)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:490)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:770)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1415)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.sql.SQLException: failed to start xa branch 172.17.0.1:8091:83249848442163200 since null
	at io.seata.rm.datasource.xa.ConnectionProxyXA.setAutoCommit(ConnectionProxyXA.java:154)
	at io.seata.rm.datasource.xa.ExecuteTemplateXA.execute(ExecuteTemplateXA.java:41)
	at io.seata.rm.datasource.xa.StatementProxyXA.executeUpdate(StatementProxyXA.java:42)
	at org.springframework.jdbc.core.JdbcTemplate$1UpdateStatementCallback.doInStatement(JdbcTemplate.java:511)
	at org.springframework.jdbc.core.JdbcTemplate$1UpdateStatementCallback.doInStatement(JdbcTemplate.java:508)
	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:376)
	... 60 more
Caused by: oracle.jdbc.xa.OracleXAException
	at oracle.jdbc.xa.OracleXAResource.checkError(OracleXAResource.java:1110)
	at oracle.jdbc.xa.client.OracleXAResource.start(OracleXAResource.java:240)
	at io.seata.rm.datasource.xa.ConnectionProxyXA.setAutoCommit(ConnectionProxyXA.java:151)
	... 65 more
2020-12-18 17:25:10.713  INFO 18272 --- [h_RMROLE_1_1_16] i.s.c.r.p.c.RmBranchRollbackProcessor    : rm handle branch rollback process:xid=172.17.0.1:8091:83249848442163200,branchId=83249850098913280,branchType=XA,resourceId=jdbc:oracle:thin:@192.168.10.53:1521:ORCL,applicationData=null
2020-12-18 17:25:10.715  INFO 18272 --- [h_RMROLE_1_1_16] io.seata.rm.AbstractRMHandler            : Branch Rollbacking: 172.17.0.1:8091:83249848442163200 83249850098913280 jdbc:oracle:thin:@192.168.10.53:1521:ORCL
2020-12-18 17:25:10.766  INFO 18272 --- [h_RMROLE_1_1_16] io.seata.rm.AbstractRMHandler            : Branch Rollbacked result: PhaseTwo_Rollbacked
2020-12-18 17:25:15.216  INFO 18272 --- [eoutChecker_1_1] i.s.c.r.netty.NettyClientChannelManager  : will connect to 192.168.66.156:8091
2020-12-18 17:25:15.219  INFO 18272 --- [eoutChecker_1_1] i.s.core.rpc.netty.NettyPoolableFactory  : NettyPool create channel to transactionRole:TMROLE,address:192.168.66.156:8091,msg:< RegisterTMRequest{applicationId='order-xa', transactionServiceGroup='my_test_tx_group'} >
2020-12-18 17:25:15.226  INFO 18272 --- [eoutChecker_1_1] i.s.c.rpc.netty.TmNettyRemotingClient    : register TM success. client version:1.5.0-SNAPSHOT, server version:1.4.0,channel:[id: 0x6aaf844a, L:/192.168.66.131:60056 - R:/192.168.66.156:8091]
2020-12-18 17:25:15.226  INFO 18272 --- [eoutChecker_1_1] i.s.core.rpc.netty.NettyPoolableFactory  : register success, cost 4 ms, version:1.4.0,role:TMROLE,channel:[id: 0x6aaf844a, L:/192.168.66.131:60056 - R:/192.168.66.156:8091]

```


### Ⅲ. Describe what you expected to happen
The transaction should success and the console not output any error.

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. config seata-xa sample's oracle db support
2. Just using browser to go to http://localhost:8084/purchase


### Ⅴ. Anything else we need to know?
   https://github.com/seata/seata/issues/2791


### Ⅵ. Environment:

- JDK version : 1.8.0_231
- OS : Windows 10 Home Edition
- Oracle: Oracle Database 11g Enterprise Edition Release 11.2.0.4.0 - 64bit Production 

@a364176773

--- a/changes/en-us/develop.md
+++ b/changes/en-us/develop.md
@@ -26,6 +26,7 @@ Add changes here for all PR submitted to the develop branch.
 - [[#5332](https://github.com/seata/seata/pull/5332)] fix bugs found in unit tests
 - [[#5413](https://github.com/seata/seata/pull/5413)] fix bad service configuration file and compilation failure
 - [[#5415](https://github.com/seata/seata/pull/5415)] fix transaction timeout on client side not execute hook and failureHandler
+- [[#5447](https://github.com/seata/seata/pull/5447)] fix oracle xa mode cannnot be used By same database
 
 
 ### optimize:

--- a/changes/zh-cn/develop.md
+++ b/changes/zh-cn/develop.md
@@ -26,6 +26,7 @@
 - [[#5332](https://github.com/seata/seata/pull/5332)] 修复单元测试中发现的bug
 - [[#5413](https://github.com/seata/seata/pull/5413)] 修复 arm64平台下的JDK和Spring兼容问题
 - [[#5415](https://github.com/seata/seata/pull/5415)] 修复客户侧事务提交前超时未执行hook和failureHandler的问题
+- [[#5447](https://github.com/seata/seata/pull/5447)] fix oracle xa mode cannnot be used By same database
 
 
 ### optimize:

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/util/SeataXAResource.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/util/SeataXAResource.java
@@ -1,3 +1,18 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
 package io.seata.rm.datasource.util;
 
 import javax.transaction.xa.XAResource;

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/util/SeataXAResource.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/util/SeataXAResource.java
@@ -1,0 +1,13 @@
+package io.seata.rm.datasource.util;
+
+import javax.transaction.xa.XAResource;
+
+/**
+ * @author PeppaO
+ * @since 2023/3/16
+ */
+public interface SeataXAResource extends XAResource {
+    // OracleXAResource Loosely Coupled Branches
+    public static final int ORATRANSLOOSE = 65536;
+
+}

--- a/rm-datasource/src/main/java/io/seata/rm/datasource/xa/ConnectionProxyXA.java
+++ b/rm-datasource/src/main/java/io/seata/rm/datasource/xa/ConnectionProxyXA.java
@@ -21,6 +21,8 @@ import javax.sql.PooledConnection;
 import javax.sql.XAConnection;
 import javax.transaction.xa.XAException;
 import javax.transaction.xa.XAResource;
+
+import com.alibaba.druid.util.JdbcConstants;
 import io.seata.common.DefaultValues;
 import io.seata.common.util.StringUtils;
 import io.seata.config.ConfigurationFactory;
@@ -30,6 +32,7 @@ import io.seata.core.model.BranchStatus;
 import io.seata.core.model.BranchType;
 import io.seata.rm.BaseDataSourceResource;
 import io.seata.rm.DefaultResourceManager;
+import io.seata.rm.datasource.util.SeataXAResource;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -62,7 +65,7 @@ public class ConnectionProxyXA extends AbstractConnectionProxyXA implements Hold
     private volatile Long prepareTime = null;
 
     private volatile Integer timeout = null;
-    
+
     private boolean shouldBeHeld = false;
 
     /**
@@ -265,7 +268,12 @@ public class ConnectionProxyXA extends AbstractConnectionProxyXA implements Hold
 
     private synchronized void start() throws XAException, SQLException {
         // 3. XA Start
-        xaResource.start(this.xaBranchXid, XAResource.TMNOFLAGS);
+        if (JdbcConstants.ORACLE.equals(resource.getDbType())) {
+            xaResource.start(this.xaBranchXid, SeataXAResource.ORATRANSLOOSE);
+        } else {
+            xaResource.start(this.xaBranchXid, XAResource.TMNOFLAGS);
+        }
+
         try {
             termination();
         } catch (SQLException e) {

